ipq806x: add support for Nokia Airscale AC400i

krusic22 opened the pull request:
Adds support for the Nokia Airscale WI4A-AC400i Access Point.
The device doesn't really exist outside of public infrastructure,
so no public page about it exist.
But can be looked up using the FCC ID: [2AD8UFZCWI4A1](https://fccid.io/2AD8UFZCWI4A1)

Hardware:
```
SoC:    Qualcomm IPQ8065
RAM:    512 MB DDR3
Flash:  256 MB NAND (Macronix MX30UF2G18AC)
        4 MB SPI-NOR (Macronix MX25U3235F)
WLAN:   Qualcomm Atheros QCA9984 - 2.4Ghz
        Qualcomm Atheros QCA9984 - 5Ghz
ETH:    eth0 - POE (100Mbps in U-Boot, 1000Mbps in OpenWrt)
        eth1 - (1000Mbps in both)
        Auto-negotiation broken on both.
USB:    USB 2.0
LED:    5G, 2.4G, ETH1, ETH2, CTRL, PWR (All support green and red)
BTN:    Reset
Other:  SD card slot (non-functional)
Serial: 115200bps, near the Ethernet transformers, labeled 9X.
        Connections from the arrow to the 9X text:
		[NC] - [TXD] - [GND] - [RXD] - [NC]
```

Install instructions are in the commit description and on the [forum](https://forum.openwrt.org/t/adding-openwrt-support-for-nokia-airscale-ac400i/78356/59?u=2in2). 

Notes:
- Auto-negotiation is broken in U-Boot and OEM firmware, so both ports are forced to run at 1000Mbps.
The POE port runs at 100Mbps in U-boot.
- The SD card slot is also non-functional (just like in the OEM firmware).
- The NAND chip is split into two partitions inside U-Boot (PartA and PartB),
we are installing OpenWrt onto PartA and leaving PartB alone, so restoring
the OEM firmware, is as easy switching the boot partition.

krusic22 committed:
ipq806x: add support for Nokia Airscale AC400i

Hardware
--------

SoC:    Qualcomm IPQ8065
RAM:    512 MB DDR3
Flash:  256 MB NAND (Macronix MX30UF2G18AC) (split into 2x128MB)
        4 MB SPI-NOR (Macronix MX25U3235F)
WLAN:   Qualcomm Atheros QCA9984 - 2.4Ghz
        Qualcomm Atheros QCA9984 - 5Ghz
ETH:    eth0 - POE (100Mbps in U-Boot, 1000Mbps in OpenWrt)
        eth1 - (1000Mbps in both)
        Auto-negotiation broken on both.
USB:    USB 2.0
LED:    5G, 2.4G, ETH1, ETH2, CTRL, PWR (All support green and red)
BTN:    Reset
Other:  SD card slot (non-functional)
Serial: 115200bps, near the Ethernet transformers, labeled 9X.
        Connections from the arrow to the 9X text:
		[NC] - [TXD] - [GND] - [RXD] - [NC]

Installation
------------

0. Connect to the device
Plug your computer into LAN2 (1000Mbps connection required).
If you use the LAN1/POE port, set your computer to force a 100Mbps link.

Connect to the device via TTL (Serial) 115200n8.
Locate the header (or solder pads) labeled 9X,
near the Ethernet jacks/transformers.
There should be an arrow on the other side of the header marking.
The connections should go like this:
(from the arrow to the 9X text): NC - TXD - GND - RXD - NC

1. Prepare for installation
While the AP is powering up, interrupt the startup process.
MAKE SURE TO CHECK YOUR CURRENT PARTITION!

If you see: "Current Partition is : partB" or
"Need to switch partition from partA to partB",
you have to force the device into partA mode, before continuing.
This can be done by changing the PKRstCnt to 5 and resetting the device.

setenv PKRstCnt 5
saveenv
reset

After you interrupt the startup process again,
you should see: Need to switch partition from partB to partA

You can now continue to the next step.

If you see: "Current Partition is : partA",
you can continue to the next step.

2. Prevent partition switching.
To prevent the device from switching partitions,
we are going to modify the startup command.
set bootcmd "setenv PKRstCnt 0; saveenv; bootipq"
setenv

3. First boot
Now, we have to boot the OpenWrt intifs.
The easiest way to do this is by using Tiny PXE.
You can also use the normal U-Boot tftp method.

Run "bootp" this will get an IP from the DHCP server
and possibly the firmware image.
If it doesn't download the firmware image, run "tftpboot".

Now run "bootm" to run the image.

You might see:
"ERROR: new format image overwritten - must RESET the board to recover"
this means that the image you are trying to load is too big.
Use a smaller image for the initial boot.

4. Install OpenWrt from initfs
Once you are booted into OpenWrt,
transfer the OpenWrt upgrade image and
use sysupgrade to install OpenWrt to the device.

Signed-off-by: Kristjan Krušič <kristjan.krusic@krusic22.com>

Ansuel commented on target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:
```diff
@@ -0,0 +1,338 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq8065.dtsi"
```
this should use the smb variant

krusic22 replied:
Marking as resolved, since I dropped 5.10 changes.

Ansuel replied:
this also apply to 5.15

krusic22 replied:
Done.

Ansuel commented on target/linux/ipq806x/image/generic.mk:
```diff
@@ -337,6 +337,20 @@ define Device/netgear_xr500
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	BOARD_NAME := ac400i
+	DEVICE_PACKAGES := -swconfig ath10k-firmware-qca9984-ct ipq-wifi-nokia-ac400i
```
Does having the swconfig package a problem? I would keep it for consistency.

krusic22 replied:
Looks like it's not required anymore. So, removed.

Ansuel commented:
Honestly I would drop the 5.10 changes since 5.15 is currently the default.

Ansuel commented on target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:
```diff
@@ -0,0 +1,327 @@
+	compatible = "nokia,ac400i", "qcom,ipq8065", "qcom,ipq8064";
+
+	reserved-memory {
+		nss@40000000 {
+			reg = <0x40000000 0x1000000>;
+			no-map;
+		};
+
+		smem: smem@41000000 {
+			reg = <0x41000000 0x200000>;
+			no-map;
+		};
+
```
```suggestion
```

Please drop these already defined in ipq8064 dtsi

krusic22 replied:
Removed.

Ansuel commented on target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:
```diff
@@ -0,0 +1,327 @@
+
+	nand@0 {
+		reg = <0>;
+		compatible = "qcom,nandcs";
```
add an extra space after this

Ansuel commented on target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:
```diff
@@ -0,0 +1,327 @@
+		compatible = "qcom,nandcs";
+		nand-ecc-strength = <4>;
+		nand-bus-width = <8>;
+		nand-ecc-step-size = <512>;
```
add an extra space after this

Ansuel commented on target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:
```diff
@@ -0,0 +1,327 @@
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
```
add an extra space after this

Ansuel commented on target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:
```diff
@@ -0,0 +1,327 @@
+			no-map;
+		};
+
+		wifi_dump@44000000 {
```
Is this really needed and used?

krusic22 replied:
Looks like it's no longer needed.

Ansuel commented:
@krusic22 we are moving the wifi board file to a separate repo... I would wait for that change to be done and then i will work on merging this. Sorry for the delay.

Ansuel commented:
@krusic22 the repo is now present. Please submit the BDF there with a pr.

krusic22 commented:
> @krusic22 the repo is now present. Please submit the BDF there with a pr.

Done: https://github.com/openwrt/firmware_qca-wireless/pull/2

Ansuel commented:
@krusic22 you added the variant naming or they were like that in the original fw? if so I would drop the -2G -5G name from them

krusic22 commented:
I added the suffix to differentiate between the 2 files.

Ansuel commented:
@krusic22 hi it's not needed since they will use the board-id so please drop them and keep the same variant. Do you want me to do the change or you can do it?

krusic22 commented:
You may go ahead and change it. Thank you.

Ansuel commented:
@krusic22 ok now you just need to bump ipq-wifi (include also you changes to the makefile) in a separate commit, give this a final test to make sure the BDF work and you are good to go

krusic22 commented:
Do I have to change anything?
I just applied the patch/PR over the latest master, used the master/ipq806x config.buildinfo,
choose the AP in the menu config and that's it.
Everything works as expected.

Ansuel commented:
@krusic22 the calibratio needs to be fixed... and as requested the ipq-wifi changes needs to be put in a separate commit. Check the prev commit of that package to understand the format of an ipq-wifi bump commit

krusic22 committed:
ipq-wifi: bump to latest git HEAD

31ff96d ipq806x: add support for Nokia Airscale AC400i
1af1df2 ath11k: ipq8074: add Netgear WAX218

Signed-off-by: Kristjan Krušič <kristjan.krusic@krusic22.com>

krusic22 commented:
> @krusic22 the calibratio needs to be fixed... and as requested the ipq-wifi changes needs to be put in a separate commit. Check the prev commit of that package to understand the format of an ipq-wifi bump commit

Done.
_I hope at least._
